# Lab | Data Cleaning and Formatting

In [1]:
import numpy as np
import pandas as pd

Objective: Apply Python programming to perform data cleaning and formatting, including identifying and handling with missing values, duplicates, outliers, errors, and using various formatting techniques such as changing data types, renaming columns, and handling string values, to prepare data for analysis.

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

In [38]:
file1 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
column_names = file1.columns
print(column_names)

cols = []
for colname in file1.columns:
    cols.append(colname.lower())

file1.columns = cols

display(file1.head())


Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='object')


customer          st gender             education customer lifetime value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly premium auto number of open complaints     policy type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle class  total claim amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they contain only valid data. For example, the gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

In [39]:
file1['gender'].nunique()
file1['gender'].unique()

array([nan, 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [40]:
gender_mapping = {
    'F': 'F',
    'M': 'M',
    'Femal': 'F',
    'Male': 'M',
    'female': 'F'
}

file1['gender'] = file1['gender'].replace(gender_mapping)
file1['gender'].unique()

array([nan, 'F', 'M'], dtype=object)

## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate. 

It is important to note that this exercise does not involve handling null values (NaN or None).

In [68]:
print(file1.dtypes)
file1.head(5)

customer                      object
st                            object
gender                        object
education                     object
customer lifetime value       object
income                       float64
monthly premium auto         float64
number of open complaints     object
policy type                   object
vehicle class                 object
total claim amount           float64
dtype: object


customer          st gender             education customer lifetime value  \
0  RB50392  Washington    nan                Master                     nan   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly premium auto number of open complaints     policy type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle class  total claim amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [70]:
column_name = 'customer lifetime value'

# Remove the ampersand sign (%)
file1[column_name] = file1[column_name].str.replace('%', '')

In [72]:
# Define the column data types
data_types = {
    'customer': str,
    'st': str,
    'gender': str,
    'education': str,
    'customer lifetime value': float,
    'income': float,
    'monthly premium auto': float,
    'number of open complaints': str,
    'policy type': str,
    'vehicle class': str,
    'total claim amount': float
}

# Update the data types of the columns
file1 = file1.astype(data_types)

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the isnull() function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (we haven't studied this yet)

Implement your chosen strategy to handle the null values. You can use the fillna() function in pandas to fill null values or dropna() function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the isnull() function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values. 


After formatting data types, as a last step, convert all the numeric variables to integers using applymap.

Save the cleaned dataset to a new CSV file.

In [74]:
file1.isna().sum()

customer                        0
st                              0
gender                          0
education                       0
customer lifetime value      2940
income                          0
monthly premium auto            0
number of open complaints       0
policy type                     0
vehicle class                   0
total claim amount              0
dtype: int64

In [77]:
len(file1)

4008

The customer lifetime value column has 2940 mising values out 4008 in total. 3/4th of the total values are missing. We could drop the column but customer lifetime value is a crutial information. To deal with the missing values, we will choose to fill null values with the mean.

In [78]:
# Fill missing values with the mean
file1['customer lifetime value'] = file1['customer lifetime value'].fillna(file1['customer lifetime value'].mean())

In [84]:
#Checking if all null values are gone
missing_values = file1['customer lifetime value'].isnull().sum()

# Print the number of missing values
print("Number of missing values in 'Customer Lifetime Value' column:", missing_values)

Number of missing values in 'Customer Lifetime Value' column: 0


In [85]:
file1.to_csv('file1_clean_ver1.csv')

## Exercise 5: Dealing with duplicates

Use the .duplicated() method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the drop_duplicates() function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using .duplicated().

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

In [90]:
duplicate_rows = file1.duplicated() #Identify duplicate rows
print(duplicate_rows)

0       False
1       False
2       False
3       False
4       False
        ...  
1067    False
1068    False
1069    False
1070    False
1071    False
Length: 1072, dtype: bool


In [92]:
#To aving having multiple rows with information about the same customer, we will choose to keep only the first occurrence of each duplicated row.
file1 = file1.drop_duplicates(keep='first')
duplicate_rows = file1.duplicated() #Verify if duplicates are handled
print(duplicate_rows)

0       False
1       False
2       False
3       False
4       False
        ...  
1067    False
1068    False
1069    False
1070    False
1071    False
Length: 1072, dtype: bool


In [93]:
file1.to_csv('file1_clean_ver2.csv')

# Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting. 

Write these functions in a separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain. 

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

In [97]:
!dir

 Volume in drive C is Local Disk
 Volume Serial Number is 9CAC-649C

 Directory of c:\Users\simra\Downloads\Ironhack\Week_1\Day_3\Afternoon

22/06/2023  23:58    <DIR>          .
15/06/2023  12:41    <DIR>          ..
15/06/2023  12:40           228.665 1.04_data_cleaning_students.ipynb
15/06/2023  12:40        29.628.769 df_final_web_data_pt_1.csv
15/06/2023  12:40            67.711 file1.csv
22/06/2023  23:50           417.545 file1_clean_ver1.csv
22/06/2023  23:58           106.329 file1_clean_ver2.csv
15/06/2023  12:40            67.745 file2.txt
15/06/2023  12:39           106.173 file3.xlsx
15/06/2023  12:39           108.362 file4.xlsx
15/06/2023  07:50            11.463 lab-dw-data-cleaning-and-formatting.ipynb
15/06/2023  17:25           324.992 merged_clean_ver1.csv
15/06/2023  17:25           309.425 merged_clean_ver2.csv
15/06/2023  17:25           309.425 merged_clean_ver3.csv
              12 File(s)     31.686.604 bytes
               2 Dir(s)  54.413.815.808 bytes free


In [116]:
!type file1.csv head 10


CONTROLN,STATE,GENDER,HV1,IC1,IC4,HVP1,IC5,POBC1,POBC2,IC2,IC3,AVGGIFT,TCODE,DOB,DOMAIN,TARGET_D
44060,FL,M,AAA896,392,520,7,21975,6,16,430,466,28,1,1901,C2,100
96093,IL,M,537.00,365,473,0,19387,1,89,415,410,5.666666667,0,0,T2,7
43333,FL,F,725.00,301,436,3,18837,11,17,340,361,4.111111111,0,2501,C2,5
21885,NC,M,AAA1095,401,413,7,14014,1,74,407,399,27.27777778,0,2208,T2,38
190108,FL,F,995.00,252,348,0,17991,5,6,280,316,6,28,0,C2,5
100640,IL,M,764.00,457,501,1,16022,2,75,477,480,25.57142857,1,6104,S2,30
18581,NC,female,775.00,318,419,5,13491,0,73,364,380,4.625,0,4407,R2,12
119038,TX,M,890.00,519,560,0,17872,2,46,525,551,6.175,1002,0,C1,7
173223,CA,F,,184,250,0,8708,2,63,206,235,8.818181818,0,1801,T2,10
157988,CA,F,AAA2294,593,617,61,17838,7,62,,619,6.666666667,28,0,S1,10
141720,NV,F,1569,673,781,22,25775,3,44,672,785,13,28,5401,T1,5
186272,CA,F,3515,521,588,97,20068,11,54,565,549,8.64,2,3401,U1,10
154301,Cali,F,1026,459,496,1,13803,12,52,470,491,11.25,2,2801,T2,10
189722,AP,Male,0,0,0,0,0


file1.csv


The system cannot find the file specified.
Error occurred while processing: head.
The system cannot find the file specified.
Error occurred while processing: 10.


In [120]:
!type file1_clean_ver1.csv head 10


,customer,st,gender,education,customer lifetime value,income,monthly premium auto,number of open complaints,policy type,vehicle class,total claim amount
0,RB50392,Washington,nan,Master,793690.2593445693,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
5,OC83172,Oregon,F,Bachelor,825629.78,62902.0,69.0,1/0/00,Personal Auto,Two-Door Car,159.383042
6,XZ87318,Oregon,F,College,538089.86,55350.0,67.0,1/0/00,Corporate Auto,Four-Door Car,321.6
7,CF85061,Arizona,M,Master,721610.03,0.0,101.0,1/0/00,Corporate Auto,Four-Door Car,363.02968
8,DY87989,Oregon,M,Bachelor,2412750.4,14072.0,71.0,1/0/00,Corporate


file1_clean_ver1.csv


The system cannot find the file specified.
Error occurred while processing: head.
The system cannot find the file specified.
Error occurred while processing: 10.


In [113]:
!type file1_clean_ver2.csv head 10

,customer,st,gender,education,customer lifetime value,income,monthly premium auto,number of open complaints,policy type,vehicle class,total claim amount
0,RB50392,Washington,nan,Master,793690.2593445693,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
5,OC83172,Oregon,F,Bachelor,825629.78,62902.0,69.0,1/0/00,Personal Auto,Two-Door Car,159.383042
6,XZ87318,Oregon,F,College,538089.86,55350.0,67.0,1/0/00,Corporate Auto,Four-Door Car,321.6
7,CF85061,Arizona,M,Master,721610.03,0.0,101.0,1/0/00,Corporate Auto,Four-Door Car,363.02968
8,DY87989,Oregon,M,Bachelor,2412750.4,14072.0,71.0,1/0/00,Corporate


file1_clean_ver2.csv


The system cannot find the file specified.
Error occurred while processing: head.
The system cannot find the file specified.
Error occurred while processing: 10.


# Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we tried to do work with that column, we wouldn't have been able to do it because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, lets get some more interesting insights!

Instructions:

- Review again the statistics for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the .describe() method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [117]:
# Review statistics for 'total claim amount' and 'customer lifetime value'
statistics = file1[['total claim amount', 'customer lifetime value']].describe()
print(statistics)


       total claim amount  customer lifetime value
count         1072.000000             1.072000e+03
mean           404.986909             7.936903e+05
std            292.890427             6.422756e+05
min              0.382107             2.004351e+05
25%            202.509051             4.054051e+05
50%            354.964564             5.888346e+05
75%            532.800000             8.937675e+05
max           2893.239678             5.816655e+06


In [123]:
# Calculate the 75th percentile of 'total claim amount'
claim_amount_threshold = file1['total claim amount'].quantile(0.75)

# Calculate the 25th percentile of 'customer lifetime value'
lifetime_value_threshold = file1['customer lifetime value'].quantile(0.25)

# Filter the DataFrame based on the thresholds
filtered_file1 = file1[(file1['total claim amount'] > claim_amount_threshold) & (file1['customer lifetime value'] < lifetime_value_threshold)]

In [124]:
# Calculate summary statistics for the filtered data
filtered_statistics = filtered_file1[['total claim amount', 'customer lifetime value']].describe()
print(filtered_statistics)

       total claim amount  customer lifetime value
count           40.000000                40.000000
mean           723.776577            341577.910000
std            149.280606             56921.472027
min            537.600000            228759.690000
25%            633.378846            304220.875000
50%            685.284189            365995.370000
75%            799.200000            387364.700000
max           1185.988301            402381.440000
